File: run/faban/master/webapps/fenxi/txt2db/mpstat.rb
# mpstat spec file
$LOAD_PATH << File.expand_path(File.dirname(__FILE__))

require 'generic_parser'

mpstat = GenericParser.new(ARGV[0], ARGV[1])
mpstat.header(/^CPU/)
mpstat.cycle(/^CPU/)
mpstat.data_type("integer")
mpstat.process
mpstat.create_table
headers = { "minf" => "Minor Faults",
  "mjf"   => "Major Faults",
  "xcal"  => "Cross calls",
  "intr"  => "Interrups",
  "ithr"  => "Interrup Threads",
  "csw"   => "Context Switches",
  "icsw"  => "Involuntary Context Switches",
  "migr"  => "Thread Migrations",
  "smtx"  => "Mutex Spins",
  "srw"   => "Reader-Writer spins",
  "syscl" => "System calls",
  "usr"   => "User Time",
  "sys"   => "System Time",
  "idl"   => "Idle Time"
}

mpstat.columns.each do |col|
  next unless headers[col]
  mpstat.section( :title => headers[col] + " per CPU", :type=>"Line", 
    :dims => 2, :query => "select Time, CPU, #{col} from TABLE")
  mpstat.section(:title =>headers[col], :summarize => col)
end

File: archive/test_executor/faban/master/webapps/fenxi/txt2db/mpstat.pl
#!/bin/perl -w

use FindBin; use lib "$FindBin::Bin"; use xanadu;

my $desc={
	'Time'=>['integer','Time'],
	'CPU'=>['varchar(8)','CPU'],
	'SET'=>['varchar(8)','Processor Set'],
	'minf'=>['integer','Minor faults'],
	'mjf'=>['integer','Major faults'],
	'xcal'=>['integer','Cross calls'],
	'intr'=>['integer','Interrupts'],
	'ithr'=>['integer','Interrupt Threads'],
	'csw'=>['integer','Context Switches'],
	'icsw'=>['integer','Involuntary Context Switches'],
	'migr'=>['integer','Thread Migrations'],
	'smtx'=>['integer','Mutex spins'],
	'srw'=>['integer','Reader-Writer spins'],
	'syscl'=>['integer','System calls'],
	'usr'=>['integer','User time'],
	'sys'=>['integer','System time'],
	'wt'=>['integer','Wait time'],
	'idl'=>['integer','Idle time'],
	'sze'=>['integer','Processor count'],
};
my $xanadu=new xanadu $desc,@ARGV;

my $interval=10;
my $i=0;
my $mp=0;
while  (<>) {
	next if (/^\s*$/);
	if (s/^\s*(SET|CPU)\s+//) {
		$set=$1;
		$i+=$interval;
		next if ($mp->{$set});
		$mp->{$set}=1;
		my @l=('Time',$set,split);
		$xanadu->create_table({
			'table' => [@l],
		});
		my $t=shift @l; shift @l; unshift @l,$t;
		$xanadu->create_view({
			"table,$set" => [@l],
		});
	} elsif (/Mpstat must always be run at some INTERVAL. Using (\d+)/) {
		$interval=$1;
	} else {
		chomp;
		$xanadu->add_row('table',$i,split);
	}
}
